File: Solutions/1863. Sum of All Subset XOR Totals/1863. Sum of All Subset XOR Totals.cpp
//brute force
class Solution {
public:
    int subsetXORSum(vector<int>& nums) {
        int res=0;
        auto subset=Allsubset(nums);
        for(const auto& s:subset){
            int tmp=0;
            for(const auto& a:s){
                tmp^=a;
            }
            res+=tmp;
        }
        return res;
    }
    vector<vector<int>> Allsubset(vector<int>& nums){
        vector<int> tmp;
        vector<vector<int>> res;
        foo(nums,res,tmp,0);
        return res;
    }
    void foo(vector<int>& nums,vector<vector<int>>& res,vector<int>& tmp,int begin){
        res.push_back(tmp);
        for(int i=begin;i<nums.size();++i){
            tmp.push_back(nums[i]);
            foo(nums,res,tmp,i+1);
            tmp.pop_back();
        }
    }
};

File: Solutions/1863. Sum of All Subset XOR Totals/1863. Sum of All Subset XOR Totals.java
//brute force
class Solution {
    public int subsetXORSum(int[] nums) {
        int res=0;
        List<List<Integer>> Allsubset=new ArrayList<>();
        List<Integer> tmp=new ArrayList<>();
        subset(nums,Allsubset,tmp,0);
        for(var s:Allsubset){
            int t=0;
            for(var n:s)
                t^=n;
            res+=t;
        }
        return res;
    }
    private void subset(int[] nums,List<List<Integer>> s,List<Integer> tmp,int begin){
        s.add(new ArrayList(tmp));
        for(int i=begin;i<nums.length;++i){
            tmp.add(nums[i]);
            subset(nums,s,tmp,i+1);
            tmp.remove(tmp.size()-1);
        }
        return;
    }
}

//trick 
class Solution {
    public int subsetXORSum(int[] nums) {
        int res=0,n=nums.length;
        for(int i=0;i<n;++i)
            res|=nums[i];
        res*=Math.pow(2,n-1);
        return res;
    }
}